### PR TITLE
Pass Pylint with a reasonable configuration

### LINKED ISSUE #2348: check-python-files.sh fails
### Description
- Type: Bug 
- Priority:  Minor

---------------------------------------------------------------
## Bug

**OS**  
linux

**mbed TLS build:**  
Version: 2.16.0
OS version: Ubuntu 16.04  

**Expected behavior**   

Running this script with pyling installed should find no error and exit zero. Additionally, when errors are found it should exit non-zero, which is not always the case.

**Actual behavior**  

On my machine it prints the following an and exits 2:
```
************* Module generate_test_code
W:282,25: Used builtin function 'map'. Using a list comprehension can be clearer. (bad-builtin)
W:298,65: Used builtin function 'map'. Using a list comprehension can be clearer. (bad-builtin)
W:399,35: Used builtin function 'map'. Using a list comprehension can be clearer. (bad-builtin)
************* Module test_generate_test_code
E: 38, 4: No name 'mock' in module 'unittest' (no-name-in-module)
E: 38, 4: Unable to import 'unittest.mock' (import-error)
************* Module mbedtls_test
E:185,22: Instance of 'MbedTlsTest' has no 'get_config_item' member (no-member)
```

On @gilles-peskine-arm machine (same OS and version!) it prints the following and exits 0 despite the error:
```
************* Module generate_test_code
W:282,25: Used builtin function 'map'. Using a list comprehension can be clearer. (bad-builtin)
W:298,65: Used builtin function 'map'. Using a list comprehension can be clearer. (bad-builtin)
W:399,35: Used builtin function 'map'. Using a list comprehension can be clearer. (bad-builtin)
************* Module test_generate_test_code
E: 38, 4: No name 'mock' in module 'unittest' (no-name-in-module)
E: 38, 4: Unable to import 'unittest.mock' (import-error)
```

**Steps to reproduce**  

Install pylint then invoke `tests/scripts/check-python-files.sh`.

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Replace the `.pylint` which had unknown modifications from the default by a completely new `.pylintrc` that only lists settings that are changed from the default.

For each issue found by Pylint, I decided whether to improve the code in such a way that the issue would no longer be present, to annotate specific instances of the issue, or to tweak the verification conditions. No message is silenced globally.

Also:

* Fix a bug whereby `check-python-files.sh` would not always report errors correctly.
* Run `pylint3`, not `pylint`, since we have Python 3 code.
* Check `scripts/*.py tests/scripts/*.py`, not just a few explicitly-named files.
* `check-python-files.sh` now runs pylint unconditionally. `all.sh` runs it if present, as before.
* In `all.sh`, don't bail out early if the crypto submodule is not present. Require the crypto submodule only if it's going to be actually used.

Fix #2348

Needs backports:

* 2.16, mbed-crypto: all except for the `all.sh` crypto submodule handling.
* 2.7: only the `check-files.py` changes, since it doesn't have `check-python-files.sh` or any other Python script.

### CODE REVIEW

File: tests/scripts/all.sh [hunk 1]
```diff
@@ -1361,6 +1361,9 @@ component_test_zeroize () {
     unset gdb_disable_aslr
 }
 
+support_check_python_files () {
+    type pylint3 >/dev/null 2>/dev/null
```
Comment (Patater) [inline tests/scripts/all.sh]: We need to make sure we have pylint3 in our CI.

Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: At the moment we have neither `pylint` nor `pylint3`. If we had `pylint`, the runs would fail.

Comment (Patater) [reply tests/scripts/all.sh]: Since you are fixing things in this PR, it'd be a shame to regress.

File: tests/scripts/all.sh [hunk 2]
```diff
@@ -222,9 +222,9 @@ cleanup()
     fi
 
     command make clean
-    cd crypto
-    command make clean
-    cd ..
+    if [ -n "$HAVE_CRYPTO_SUBMODULE" ]; then
+        command make -C crypto clean
```
Comment (mpg) [inline tests/scripts/all.sh]: This conflicts with https://github.com/ARMmbed/mbedtls/pull/2438 which has similar fixes. Also, I fail to see the connection with the main purpose of this PR (address Pylint issues). Please consider removing this commit from the PR in the interest of having single-purpose PRs.

File: tests/scripts/all.sh [hunk 3]
```diff
@@ -555,6 +552,26 @@ pre_check_tools () {
     "$@" scripts/output_env.sh
 }
 
+need_submodule () {
+    case $RUN_COMPONENTS in
+        *test_submodule*) return 0;;
```
Comment (mpg) [inline tests/scripts/all.sh]: This is not correct, as `component_test_use_psa_crypto_full_cmake_asan` relies on the submodule as well.

Also, in the near future (#2403 and hopefully soon after that, de-duplicating files) all tests will require the submodule, so do we really want to invest time in making `all.sh` working without it now?

File: .pylintrc [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+method-rgx=[a-z_][a-z0-9_]{2,35}$
+
+# Allow module names containing a dash (but no underscore or uppercase letter).
+# They are whole programs, not meant to be included by another module.
+# [invalid-name]
+module-rgx=(([a-z_][a-z0-9_]*)|([A-Z][a-zA-Z0-9]+)|[a-z][-0-9a-z]+)$
+
+# Some functions don't need docstrings.
+# [missing-docstring]
+no-docstring-rgx=(run_)main$
```
Comment (mpg) [inline .pylintrc]: I'm not sure I understand this regex: do you mean `(run_)?main$`?

File: tests/scripts/all.sh (hunk 2 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: This is the third time I've been annoyed by having to check out the submodule even though it was irrelevant to what I was doing, hence I got rid of the submodule obligation. But ok, I'll remove it from this PR.

File: .pylintrc (hunk 4 above)
Comment (gilles-peskine-arm) [reply .pylintrc]: Indeed. FIxed.

**Comment by gilles-peskine-arm** [comment]
I rebased this PR to remove the commit that improves submodule handling in `all.sh`. I made a new commit to fix the [typo in `no-docstring-rgx`](https://github.com/ARMmbed/mbedtls/pull/2464#discussion_r260644068).

**Comment by Patater** [comment]
@mpg's comments have been addressed. CI failure is FreeBSD timing test, a known issue. Two approved reviewers. Ready to merge.

**Pull request merged**